biz.aQute.bndlib/src/aQute/bnd/cdi/AnnotationReader.java

```diff
@@ -0,0 +1,340 @@
+import aQute.bnd.osgi.Instruction;
+import aQute.bnd.signatures.ClassTypeSignature;
+import aQute.bnd.signatures.FieldSignature;
+import aQute.bnd.signatures.JavaTypeSignature;
+import aQute.bnd.signatures.MethodSignature;
+import aQute.bnd.signatures.TypeArgument;
+import aQute.bnd.version.Version;
+
+public class AnnotationReader extends ClassDataCollector {
+	public static final Version			V1_0						= new Version("1.0.0");								// "1.0.0"
+
+	private static final Instruction	COMPONENTSCOPED_INSTR	= new Instruction(
+		"org.osgi.service.cdi.annotations.ComponentScoped");
+	private static final Instruction	DEPENDENT_INSTR		= new Instruction("javax.enterprise.context.Dependent");
+	private static final Instruction	EXTENSION_INSTR		= new Instruction("javax.enterprise.inject.spi.Extension");
+	private static final Instruction	INTERCEPTOR_INSTR	= new Instruction(
+		"javax.interceptor.Interceptor");
+	private static final Instruction	NORMALSCOPE_INSTR			= new Instruction(
+		"javax.enterprise.context.NormalScope");
+	private static final Instruction	STEREOTYPE_INSTR			= new Instruction(
+		"javax.enterprise.inject.Stereotype");
+	private static final Instruction	VETOED_INSTR		= new Instruction("javax.enterprise.inject.Vetoed");
+
+	final Analyzer						analyzer;
+	final Clazz							clazz;
+	final EnumSet<Options>				options;
+	final Map<PackageRef, PackageDef>	packageInfos			= new HashMap<>();
+	boolean								baseclass				= true;
+	final List<BeanDef>					definitions				= new ArrayList<>();
+	TypeRef								extendsClass;
+	TypeRef								interfaces[];
+	FieldDef							member;
+	int									minimumCardinality		= -1;
+	ReferenceDef						referenceDef;
+
+	AnnotationReader(Analyzer analyzer, Clazz clazz, EnumSet<Options> options) {
+		this.analyzer = requireNonNull(analyzer);
+		this.clazz = clazz;
+		this.options = options;
+		this.definitions.add(new BeanDef());
+	}
+
+	public static List<BeanDef> getDefinition(Clazz c, Analyzer analyzer, EnumSet<Options> options) throws Exception {
+		AnnotationReader r = new AnnotationReader(analyzer, c, options);
+		return r.getDefs();
+	}
+
+	private List<BeanDef> getDefs() throws Exception {
+		if (clazz.isEnum() || clazz.isInterface() || clazz.isAnnotation()) {
+			// These types cannot be managed beans so don't bother scanning
+			// them. See
+			// http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#what_classes_are_beans
+			return null;
+		}
+
+		// if discovery mode is 'all', all concrete classes are considered
+		if (options.contains(Options.all) && !clazz.is(QUERY.CONCRETE, null, analyzer)) {
+			return null;
+		}
+
+		// if discovery mode is 'annotated', only classes annotated with a bean
+		// defining annotation are considered. See
+		// http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#bean_defining_annotations
+		if (options.contains(Options.annotated)) {
+			if (clazz.is(QUERY.ANNOTATED, VETOED_INSTR, analyzer)
+				|| !clazz.is(QUERY.INDIRECTLY_ANNOTATED, COMPONENTSCOPED_INSTR, analyzer)
```

Comment:
I know there are precedence rules for the `||` and `&&` operations in this sequence. But it would be more clear and explicit to use parenthesis liberally here to avoid confusion and subtle future bugs.